include/swift/Sema/CSFix.h

```diff
@@ -409,14 +409,14 @@ class ConstraintFix {
   FixKind Kind;
   ConstraintLocator *Locator;
 
-  /// Determines whether this fix is simplify a warning which doesn't
-  /// require immediate source changes.
-  bool IsWarning;
+  /// The behavior limit to apply to the diagnostics emitted.
+  DiagnosticBehavior behaviorLimit;
 
 public:
   ConstraintFix(ConstraintSystem &cs, FixKind kind, ConstraintLocator *locator,
-                bool warning = false)
-      : CS(cs), Kind(kind), Locator(locator), IsWarning(warning) {}
```

Comment:
That not the way `warning` is used now. When `warning` evaluates true, we would call `limitBehavior` on the diagnostic to downgrade the diagnostic to a warning, so it's guaranteed to be emitted as a warning. I've generalized this to allow us to specify the behavior limit in more detail: `Unspecified` means keep it an error, `Warning` to make it a warning, or `Ignore` to suppress the diagnostic entirely. 

There is a remaining failure case here, where the behavior limit is `Unspecified` but the actual diagnostic is defined as a warning in the diagnostics file. We could certainly do a pass through the diagnostics emitted from CSDiagnostics to make sure none of them are warnings---and for any that are warnings, convert them to errors with a `Warning` behavior limit.

Reply:
That’s why I said - more and more changes erode original meaning of warning indicator…

Reply:
To re-phrase - the original intention behind the warning bit was to indicate to the solver that the fix has no impact on the solution and should produce a warning when diagnosed, but then it was changed to impact diagnostics as well by changing how `emitDiagnostic` behaves. My suggestion would be to have a warning and behavior indicators to indicate the difference of expectation.

Reply:
The original warning indicator required careful synchronization between the declaration of the diagnostic and the creation of the failure instance. That's no longer necessary: all of these can be declared as errors and then downgraded to warnings via this mechanism. The new definition is less error-prone and more expressive.

Reply:
It wasn't a bug as you make it sound - it was a feature to make sure that the diagnostic produced by the FailureDiagnostic was indeed a warning the fix expected because if it wasn't it creates other considerations - whether the fix should have been marked a warning in the first place. I'm not against limiting behavior and new flag affecting `isWarning()` result but I'm trying to convey that the warning flag use is completely different.

Reply:
The warning flag was originally about determining whether we can proceed to constraint application or not, and should only be set when we can still create a proper expression. Are you concerned about the case where we want to produce an error but still proceed to constraint application?

Reply:
It isn't limited to constraint application - it also determines whether the score gets adjusted or not. The idea was to set the flag on the fix and then double-check that the diagnostic was indeed a warning because otherwise it's a bug in the solver logic which created a warning instead of an error, so either the fix has to be changed or the diagnostic itself has to be adjusted. I'm concerned about a couple of things here - fix is a warning instead of an error that that leads to ill-formed type-checked AST that crashes in SILGen, diagnostic as an error although it actually should always be a warning but is marked as an error that that is never noticed due to the downgrades. I think there should be a clear distinction - this fix is a warning vs. this fix is actually an error but behaves as a warning in some conditions.

Reply:
I see. And fixes that are errors-downgraded-to-warnings should count in the score, whereas always-warnings should not?

Reply:
Yes, I think that would be correct solver behavior because downgrade could lead to ambiguities or picking incorrect choice and lifting the restriction would lead to change in overloading behavior for existing code. Also fixes are not diagnostics so IMHO it would be a good idea to add a new enum which represents a downgrade that would only include relevant choices - `Warning` and `Ignore` (`Ignore` - we need to figure out what that even means in terms of fixes) and would then get translated into a `DiagnosticBehavior` for `FailureDiangostic`.

Reply:
This is handled via https://github.com/apple/swift/pull/59958